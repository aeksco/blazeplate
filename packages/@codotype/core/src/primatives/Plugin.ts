import {
    PluginMetadata,
    SchemaEditorConfiguration,
    ExperienceRecommendation,
    PluginCreator,
    ConfigurationGroup,
} from "..";
import { SchemaEditorBuilder } from "./SchemaEditor";
import { ProjectInput } from "../project";

interface PluginBuilderParams {
    id: string;
    label: string;
    description: string;
    project_path: string;
    version?: string;
    codotypeVersion?: string;
    icon?: string;
    homepage?: string;
    techTags?: string[];
    typeTags?: string[];
    documentation?: string;
    experience?: ExperienceRecommendation;
    configurationGroups?: ConfigurationGroup[];
    schemaEditorConfiguration?: SchemaEditorConfiguration;
    createdBy?: PluginCreator;
    exampleProjects?: ProjectInput[];
}

export class PluginBuilder implements PluginMetadata {
    id: string;
    label: string;
    description: string;
    project_path: string;
    version: string = "0.0.1";
    codotypeVersion: string = "0.0.0";
    icon: string = "";
    homepage: string = "";
    techTags: string[] = [];
    typeTags: string[] = [];
    documentation: string = "";
    experience: ExperienceRecommendation = ExperienceRecommendation.BEGINNER;
    configurationGroups: ConfigurationGroup[] = [];
    schemaEditorConfiguration: SchemaEditorConfiguration = new SchemaEditorBuilder(
        {
            supportedDatatypes: [],
            supportedRelations: [],
        },
    );
    createdBy: PluginCreator = {
        name: "",
        contact: {},
    };
    exampleProjects: ProjectInput[] = [];

    constructor(params: PluginBuilderParams) {
        this.id = params.id;
        this.label = params.label;
        this.description = params.description;
        this.project_path = params.project_path;

        this.version = params.version || this.version;
        this.codotypeVersion = params.codotypeVersion || this.codotypeVersion;
        this.icon = params.icon || this.icon;
        this.homepage = params.homepage || this.homepage;
        this.techTags = params.techTags || this.techTags;
        this.typeTags = params.typeTags || this.typeTags;
        this.documentation = params.documentation || this.documentation;
        this.experience = params.experience || this.experience;
        this.configurationGroups =
            params.configurationGroups || this.configurationGroups;
        this.schemaEditorConfiguration =
            params.schemaEditorConfiguration || this.schemaEditorConfiguration;
        this.createdBy = params.createdBy || this.createdBy;
        this.exampleProjects = params.exampleProjects || this.exampleProjects;
    }
}
