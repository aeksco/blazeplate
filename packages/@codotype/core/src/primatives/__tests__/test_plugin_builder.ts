import { PluginBuilder } from "../Plugin";

// // // //

describe("factory", () => {
    test("works", () => {
        const plugin = new PluginBuilder({
            id: "my-codotype-plugin",
            label: "My Codotype Plugin",
            description: "My Codotype Plugin description",
            project_path: "my_codotype_plugin",
        });

        expect(plugin).toMatchSnapshot();
    });
});
