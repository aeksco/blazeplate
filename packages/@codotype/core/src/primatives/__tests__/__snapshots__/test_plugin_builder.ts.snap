// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
PluginBuilder {
  "codotypeVersion": "0.0.0",
  "configurationGroups": Array [],
  "createdBy": Object {
    "contact": Object {},
    "name": "",
  },
  "description": "My Codotype Plugin description",
  "documentation": "",
  "exampleProjects": Array [],
  "experience": "beginner",
  "homepage": "",
  "icon": "",
  "id": "my-codotype-plugin",
  "label": "My Codotype Plugin",
  "project_path": "my_codotype_plugin",
  "schemaEditorConfiguration": SchemaEditorBuilder {
    "attributeAddons": Array [],
    "configurationGroups": Array [],
    "defaultAttributes": Array [],
    "defaultRelations": Array [],
    "defaultSchemas": Array [],
    "documentation": "",
    "enableAttributeDefaultValue": false,
    "supportedDatatypes": Array [],
    "supportedRelations": Array [],
  },
  "techTags": Array [],
  "typeTags": Array [],
  "version": "0.0.1",
}
`;
